### PR TITLE
Flatten nested ternaries

### LINKED ISSUE #737: alternate ternary formatting
Here is how Prettier currently reformats nested ternaries:
````
const value = condition1
  ? value1
  : condition2
      ? value2
      : condition3
          ? value3
          : value4;
````
This isn't so bad when there are only two ternaries, but I sometimes have many more because I essentially use them as an alternative to a switch statement when each case only gets a single value.

Here is how I prefer to format this case:
````
const value =
  condition1 ? value1 :
  condition2 ? value2 :
  condition3 ? value3 :
  value4;
````
I like this because it clearly associates each condition with its corresponding value and the indentation doesn't increase regardless of the number of ternaries.

Perhaps some like the current formatting because it discourages use of nested ternaries, but I think my example shows that you can have any number without making the code hard to read.

Would you consider adding an option to support this alternate formatting? Maybe something like this? `--ternary=condval`

### PR DESCRIPTION
**Pull request by vjeux**
This keeps coming up that the way prettier prints nested ternaries is not good. Here's an attempt at solving the situation. A good compromise is to
- keep printing ternaries the way they are but do not indent nested ones
- have a single group for chained ternaries, this way the last ones also break instead of awkwardly being fitting in a single line

The implementation is pretty straightforward, just look at the parent and print it differently if it's a ternary and we are in the second group.

Before:
```js
const value = condition1
  ? value1
  : condition2 ? value2 : condition3 ? value3 : value4;
```

After:
```js
const value = condition1
  ? value1
  : condition2
  ? value2
  : condition3
  ? value3
  : value4;
```

Fixes #737

### CODE REVIEW

**Comment by mvolkmann** [comment]
This is SO much better than the current formatting.

**Comment by SimenB** [comment]
Why not indent nested?

**Comment by mvolkmann** [comment]
I'm confused, @SimenB. It indents nested ternaries now. Each one is indented more than the previous one. If there is more than one ternary, it looks like code I've never seen a person write. It becomes deeply nested fast. I think it's fair to say that a side goal of the current formatting is to discourage use of nested ternaries. I think the main issue is whether that is a good goal. Me and several others think it is possible to format nested ternaries in a way that looks nice and is not confusing.

**Comment by SimenB** [comment]
I find 
```js
const value = condition1
  ? value1
  : condition2
    ? value2
    : condition3
      ? value3
      : value4;
```
to be easier to follow.

Of course, I haven't written a nested ternary for a couple of years (eslint would tell me off), so this doesn't really impact me 😄

**Comment by mvolkmann** [comment]
That's not exactly how Prettier does it though. It indents more than that. Many lines get indented four spaces instead of two. Your example looks fine for three ternaries, but not for say seven.

**Comment by SimenB** [comment]
7 nested ternaries should really be refactored anyways, that's really unreadable :P

**Comment by mvolkmann** [comment]
I just don't understand how anyone could claim it is unreadable when written like this:
```js
const value =
  condition1 ? value1 :
  condition2 ? value2 :
  condition3 ? value3 :
  condition4 ? value4 :
  condition5 ? value5 :
  condition6 ? value6 :
  condition7 ? value7 :
  value8;
```

But everyone is entitled to their opinion and I've already lost this battle.

**Comment by jlongster** [comment]
One of the things to think about here is how it looks when the expression (between `:` or `?`) is broken up across multiple lines. If we're going to change this (and I think we should) it might be nice to go ahead and explore this syntax as well:

```
foo ? (
  bigFunctionCall(
    arg1,
    arg2
  )
) : bar ? (
  // some other big expr
) : baz ? (
  // something else
)
```

Or some variant, where we use parentheses to aid making it more structured since the expressions are bigger.

This could possible work with @mvolkmann's syntax, but we move the value inside parens if it breaks across multiple lines.

**Comment by vjeux** [comment]
@jlongster would you print it as

```js
foo ? (
  bigFunctionCall(
    arg1,
    arg2
  )
) : bar
```

for a single ternary in order to be consistent?

**Comment by timdorr** [comment]
I think what would be helpful to the discussion would be to avoid using easily-aligned examples. Make it messy so you're not picking something because the example case lines up nicely. Here's one way to make it look not so pretty, so you're working from the practical case:
```js
const avatarURL = user.avatar && user.avatar.url
  ? user.avatar.url
  : user.name !== null
  ? generateAvatar(
    user.name.first,
    user.name.last
  )
  : generateAvatar(user.id)
```
That might be one way it comes out.

Personally, I find that a little harder to read than other forms. It's better than the current behavior, but I think it can be improved with indentation. Here's @SimenB's case:
```js
const avatarURL = user.avatar && user.avatar.url
  ? user.avatar.url
  : user.name !== null
    ? generateAvatar(
      user.name.first,
      user.name.last
    )
    : generateAvatar(user.id)
```
And @mvolkmann's:
```js
const avatarURL = 
  user.avatar && user.avatar.url ? user.avatar.url : 
  user.name !== null ? generateAvatar(
      user.name.first,
      user.name.last
    ) : generateAvatar(user.id)
```
Thoughts?

**Comment by SimenB** [comment]
Out of those, I find  the middle one the most readable. But again, I haven't read nested ternaries in a while, as airbnb style enforces no nesting: https://github.com/airbnb/javascript/blob/master/README.md#comparison--nested-ternaries

**Comment by kjbekkelund** [comment]
Agreed on the middle being the most readable.

**Comment by ianstormtaylor** [comment]
I agree with the middle syntax there, that's how I would write those to be most readable later. (Although I agree, I'd avoid it generally.)

**Comment by mvolkmann** [comment]
Good idea to consider a more complicated case! Here is how I would write it:
```js
const avatarURL = 
  user.avatar && user.avatar.url ? user.avatar.url : 
  user.name !== null ?
    generateAvatar(
      user.name.first,
      user.name.last
    ) :
  generateAvatar(user.id)
```
The rule here would be that a value can only be on the same line as its condition if it fits within 80 columns. Otherwise the whole value is moved to the next line and indented.

The only thing I don't prefer about the @SimenB's case is that each ternary gets nested deeper than than the one before. That's the primary thing I'd like to avoid. The first example and mine do avoid that.

**Comment by ForbesLindesay** [comment]
I'd write:

```js
const avatarURL = user.avatar && user.avatar.url
  ? user.avatar.url
  : user.name !== null
  ? generateAvatar(
      user.name.first,
      user.name.last
    )
  : generateAvatar(user.id);
```

Or

```js
const avatarURL = (
  user.avatar && user.avatar.url
  ? user.avatar.url
  : user.name !== null
  ? generateAvatar(
      user.name.first,
      user.name.last
    )
  : generateAvatar(user.id)
);
```

I prefer the latter but I can accept the former.

**Comment by tlrobinson** [comment]
We use this pattern a lot in JSX:

```javascript
<div>
  { conditionA ?
    <div>foo</div>
  : conditionB ?
    <div>
      <div>bar</div>
    </div>
  : null }
</div>
```

This would be ok (but not as readable IMHO):

```javascript
<div>
  { conditionA
  ? <div>foo</div>
  : conditionB
  ? <div>foo</div>
  : null }
</div>
```

...but would start looking weird with multilines.

I guess something like this would be ok:

```javascript
<div>
  { conditionA
  ? <div>foo</div>
  : conditionB
  ?
    <div>
      <div>bar</div>
    </div>
  : null }
</div>
```

Just my 2¢!

BTW, this is the only thing holding us back from prettierifying our entire codebase, awesome tool otherwise!

**Comment by ianstormtaylor** [comment]
The JSX case is actually interesting, because it's (one of?) the only places where the ternary is required to keep things inline easily, and where you actually might not want nesting because it adds confusion to the meaning of the output. But I'm not sure personally that it feels like a common enough case to merit weird behavior.

Otherwise, for all of the pure JS cases, I think not nesting is purely a stylistic choice that should be ignored in favor of being consistent with nesting in other places. (Often just a "what, why?!" choice.)

(That said: I don't actually care what style Prettier chooses for these kinds of nested ternaries because they feel like huge edge cases to me, so please ignore everything I say!)

**Comment by kentor** [comment]
I'd like the formatting to parallel `if/else` statements

Suppose `if`s were expressions instead of statements:
```jsx
<div>
  {if (condition) {
    <A />
  } else if (condition2) {
    <B />
  } else {
    null
  }}
</div>
```

With ternaries this would be
```jsx
<div>
  {condition ?
    <A />
  : condition2 ?
    <B />
  :
    null
  }
</div>
```

That is the most readable to me. This is similar to @ryanflorence's style but without the unnecessary parentheses.

This is also holding me back from adopting prettier at the moment.

**Comment by bakkot** [comment]
@kentor, there's an important difference from `if/else`, which is that it's not obvious whether the thing after `:` is a condition or result until the *end* of the line (or the next line). I think the increasing-indent style (@SimonB's) is the most readable, for this reason.

I also feel the increasing-indent style is the most consistent with how prettier formats expressions more generally, which seems like it ought to be a guiding concern for edge cases like this.

**Comment by megamaster85** [comment]
Help

**Comment by umidbekkarimov** [comment]
The way prettier formats it made me turn off eslint rules about ternaries. It's awesome now!

```javascript
<div>
  {props.dirty
    ? <FlatButton label="Reset" onTouchTap={props.reset} />
        ? props.onDismiss
        : <FlatButton label="Dismiss" onTouchTap={props.onDismiss} />
    : null}

  <FlatButton label="Submit" type="submit" />
</div>;
```

If ternaries will look like this, I will have to refactor everything back to `&&` style.

```javascript
<div>
  {props.dirty
    ? <FlatButton label="Reset" onTouchTap={props.reset} />
    ? props.onDismiss
    : <FlatButton label="Dismiss" onTouchTap={props.onDismiss} />
    : null}

  <FlatButton label="Submit" type="submit" />
</div>;
```

**Comment by mvolkmann** [comment]
I'm having trouble mentally parsing this example. It has two consecutive ?'s followed by two consecutive :'s. Typically those alternate. How does it make sense for a ? to follow FlatButton JSX?

In any case, increasing the indents for more than three ternaries really discourages the use of more that three. I supposed for some people that is a goal, but it isn't mine.

---
R. Mark Volkmann
Object Computing, Inc.

**Comment by umidbekkarimov** [comment]
> consecutive :'s. Typically those alternate. How does it make sense for a ? to follow FlatButton JSX?

Sorry @mvolkmann, I gave bad example there (i'm not good in ternaries) and removed this comment but forgot to provide a better example: 

How it looks now:

```javascript
return !props.id
  ? <div>Invalid ID: {props.id}</div>
  : props.fetching
      ? <div>Loading ...</div>
      : props.notFound
          ? <div>
              Not Found: {props.id}
            </div>
          : <div>{props.name}</div>;
```

In this case readability would be same, but only with one line conditions:

```javascript
return
  !props.id ? <div>Invalid ID: {props.id}</div> :
  props.fetching ? <div>Loading ...</div> :
  props.notFound ? (
      <div>Not Found: {props.id}</div>
  ) : <div>{props.name}</div>;
```

But this one can confuse.

```javascript
return !props.id
  ? <div>Invalid ID: {props.id}</div>
  : props.fetching
  ? <div>Loading ...</div>
  : props.notFound
  ? <div>
      Not Found: {props.id}
    </div>
  : <div>{props.name}</div>;
```

**Comment by nmn** [comment]
1. Prettier needs to handle nested ternaries regardless of developer's opinion of whether they should be used. They are also pretty common within JSX because of the lack of an if expression.
2. The original solution by @vjeux works great.
3. It's also the formatting championed by `Standard` and it's many variants.
4. Multiple indentation is less scalable, as you could easily run over the column character limit.

In Conclusion,
Let's Merge This!

**Comment by azz** [comment]
@kentor I hope [`do { }`](https://gist.github.com/dherman/1c97dfb25179fa34a41b5fff040f9879) makes it into the spec so we can forget that nested ternaries are a thing 😞. 

```ts
<div>
  {do { 
    if (condition) {
      <A />
    } else if (condition2) {
      <B />
    } else {
      null
    }
  }}
</div>
```

**Comment by vjeux** [comment]
I'm doing a cleanup of all the pull requests, I'm going to close this one. We still want to figure out what to do with nested ternaries though, so the discussion is not over yet.

**Comment by reywright** [comment]
I'm a big fan of a lot of these. I think anything is a huge improvement over how ternaries format now. A lot of good suggestions in this thread.
